airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DestinationV1V2Migrator.java

```diff
@@ -0,0 +1,171 @@
+
+import static io.airbyte.integrations.base.destination.typing_deduping.Constants.RAW_TABLE_EXPECTED_V1_COLUMNS;
+import static io.airbyte.integrations.base.destination.typing_deduping.Constants.RAW_TABLE_EXPECTED_V2_COLUMNS;
+
+import io.airbyte.protocol.models.AirbyteStreamNameNamespacePair;
+import io.airbyte.protocol.models.v0.DestinationSyncMode;
+import io.airbyte.protocol.models.v0.SyncMode;
+import java.util.Collection;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public interface DestinationV1V2Migrator<DialectTableDefinition> {
+
+    Logger LOGGER = LoggerFactory.getLogger(DestinationV1V2Migrator.class);
+
+    /**
+     * This is the primary entrypoint to this interface
+     * <p>
+     * Determine whether a migration is necessary for a given stream and if so, migrate the raw table and rebuild the final table with a soft reset
+     *
+     * @param sqlGenerator       the class to use to generate sql
+     * @param destinationHandler the handler to execute the sql statements
+     * @param streamConfig       the stream to assess migration needs
+     * @return an optional MigrationResult if a migration occurred
+     */
+    default Optional<MigrationResult> migrateIfNecessary(
+        final SqlGenerator<DialectTableDefinition> sqlGenerator,
+        final DestinationHandler<DialectTableDefinition> destinationHandler,
+        final StreamConfig streamConfig
+    ) {
+        if (shouldMigrate(streamConfig)) {
+            LOGGER.info("Starting v2 Migration for stream {}", streamConfig.id().finalName());
+            return Optional.of(migrate(sqlGenerator, destinationHandler, streamConfig));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Determine whether a given stream needs to be migrated from v1 to v2
+     *
+     * @param streamConfig the stream in question
+     * @return whether to migrate the stream
+     */
+    default boolean shouldMigrate(final StreamConfig streamConfig) {
+        return isMigrationRequiredForSyncMode(streamConfig.syncMode(), streamConfig.destinationSyncMode())
+            && !doesValidV2RawTableAlreadyExist(streamConfig)
+            && doesValidV1RawTableExist(convertToV1RawName(streamConfig));
+    }
+
+    /**
+     * Execute sql statements that converts a v1 raw table to a v2 raw table. Leaves the v1 raw table intact
+     *
+     * @param sqlGenerator       the class which generates dialect specific sql statements
+     * @param destinationHandler the class which executes the sql statements
+     * @param streamConfig       the stream to migrate the raw table of
+     * @return a successful migration result
+     */
+    default MigrationResult migrate(final SqlGenerator<DialectTableDefinition> sqlGenerator,
+        final DestinationHandler<DialectTableDefinition> destinationHandler,
+        final StreamConfig streamConfig) {
+        final var migrateAndReset = String.join("\n",
```

Comment:
I'm sure anything is possible. The migration statement cannot be wrapped in a transaction (big query yelled at me) its just a create table from a select statement. I think if it were to fail for an unexpected reason we could retry it on the next attempt, we're not removing the old raw table. The soft reset I think is wrapped by a transaction. I think its also probably not crucial to do the soft reset here since the schema may have changed from v1 to v2 and this would also trigger a soft reset on a normal write, so i think if the migration succeeds but the soft reset fails we should be fine? (assuming the issue is not with the soft reset logic)

Reply:
sounds good to me. and I guess in the failure case, we're not touching the raw tables so users can easily rollback + rebuild their v1 tables :tu: